File: backend/java/elo-calculator-backend/src/main/java/de/iits/elocalculatorbackend/ranking/service/RankingService.java
package de.iits.elocalculatorbackend.ranking.service;

import de.iits.elocalculatorbackend.player.model.dto.PlayerResponseDto;
import de.iits.elocalculatorbackend.player.repository.PlayerRepository;
import de.iits.elocalculatorbackend.ranking.model.dto.RankingResponseDto;
import lombok.AllArgsConstructor;
import org.springframework.data.domain.Limit;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.IntStream;

@Service
@AllArgsConstructor
public class RankingService {

    private PlayerRepository playerRepository;

    public List<RankingResponseDto> getTopPlayers(int top) {
        var players = playerRepository.
                findByOrderByEloDesc(Limit.of(top)).stream()
                .map(PlayerResponseDto::new)
                .toList();

        return IntStream.range(0, players.size())
                .mapToObj(index -> new RankingResponseDto(index + 1, players.get(index)))
                .toList();
    }
}

File: backend/springboot-kotlin/src/main/kotlin/de/iits/elo/ranking/service/RankingService.kt
package de.iits.elo.ranking.service

import de.iits.elo.player.repository.PlayerRepository
import de.iits.elo.ranking.model.dto.RankingResponseDto
import org.springframework.data.domain.PageRequest
import org.springframework.stereotype.Service

@Service
class RankingService(
    private val playerRepository: PlayerRepository,
) {

    fun getTopPlayers(number: Int) =
        playerRepository.findByOrderByEloDesc(PageRequest.of(0, number))
            .mapIndexed { index, user -> RankingResponseDto(rank = index + 1, player = user) }
}
